### PR TITLE
refactor: move rate limiter to zetaclient and add metrics

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

- [x] move rate limiter logic to `zetaclient`. The method is named as `ApplyRateLimiter`
- [x] leave query part in zetacore to be performant. The method is named as `RateLimiterInput`.
- [x] order pending cctxs by height across all chains to be **first-come-first-serve**
- [x] add metrics of `percentage_of_rate_reached` to monitor current rate.

Closes: [2074](https://github.com/zeta-chain/node/issues/2074)
Closes: [2082](https://github.com/zeta-chain/node/issues/2082)
Closes: [2092](https://github.com/zeta-chain/node/issues/2092)
Closes: [2093](https://github.com/zeta-chain/node/issues/2093)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: cmd/zetae2e/local/local.go [hunk 1]
```diff
@@ -313,6 +313,7 @@ func localE2ETest(cmd *cobra.Command, _ []string) {
 			e2etests.TestUpdateBytecodeZRC20Name,
 			e2etests.TestUpdateBytecodeConnectorName,
 			e2etests.TestDepositEtherLiquidityCapName,
+			e2etests.TestRateLimiterName,
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetae2e/local/local.go:316]: good catch. It probably came from my git merge.

File: zetaclient/ratelimiter/rate_limiter.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
+package ratelimiter
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/ratelimiter/rate_limiter.go:1]: The linter will fail with `don't use an underscore in package name`

File: cmd/zetae2e/local/local.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ws4charlie) [reply cmd/zetae2e/local/local.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: x/crosschain/keeper/grpc_query_cctx_rate_limit.go [hunk 3]
```diff
@@ -241,52 +393,46 @@ func (k Keeper) ListPendingCctxWithinRateLimit(c context.Context, req *types.Que
 func ConvertCctxValue(
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx_rate_limit.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: x/crosschain/keeper/grpc_query_cctx_rate_limit.go [hunk 4]
```diff
@@ -302,13 +448,12 @@ func ConvertCctxValue(
 func rateLimitExceeded(
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx_rate_limit.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: x/crosschain/keeper/rate_limiter_flags.go [hunk 5]
```diff
@@ -46,16 +43,37 @@ func (k Keeper) GetRateLimiterRates(ctx sdk.Context) (map[int64]sdk.Dec, map[int
...
+			Rate:     conversion.Rate,
+		})
+	}
+	return assetRateList
+}
+
+// BuildAssetRateMapFromList builds maps (foreign chain id -> asset -> rate) from a list of gas and erc20 asset rates
+// The 1st map: foreign chain id -> gas coin asset rate
+// The 2nd map: foreign chain id -> erc20 asset -> erc20 coin asset rate
+func BuildAssetRateMapFromList(assetRates []*types.AssetRate) (map[int64]*types.AssetRate, map[int64]map[string]*types.AssetRate) {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/rate_limiter_flags.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: x/crosschain/keeper/rate_limiter_flags.go [hunk 6]
```diff
@@ -29,15 +29,12 @@ func (k Keeper) GetRateLimiterFlags(ctx sdk.Context) (val types.RateLimiterFlags
 	return val, true
 }
 
-// GetRateLimiterRates returns two maps of foreign coins and their rates
-// The 1st map: foreign chain id -> gas coin rate
-// The 2nd map: foreign chain id -> erc20 asset -> erc20 coin rate
-func (k Keeper) GetRateLimiterRates(ctx sdk.Context) (map[int64]sdk.Dec, map[int64]map[string]sdk.Dec) {
+// GetRateLimiterAssetRateList returns a list of all foreign asset rate
+func (k Keeper) GetRateLimiterAssetRateList(ctx sdk.Context) []*types.AssetRate {
 	rateLimiterFlags, _ := k.GetRateLimiterFlags(ctx)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/rate_limiter_flags.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: zetaclient/core_context/zeta_core_context.go [hunk 7]
```diff
@@ -83,6 +83,18 @@ func (c *ZetaCoreContext) GetEnabledChains() []chains.Chain {
 	return copiedChains
 }
 
+func (c *ZetaCoreContext) GetEnabledForeignChains() []chains.Chain {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/core_context/zeta_core_context.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/core_context/zeta_core_context.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: zetaclient/core_context/zeta_core_context.go [hunk 8]
```diff
@@ -83,6 +83,18 @@ func (c *ZetaCoreContext) GetEnabledChains() []chains.Chain {
...
+	c.coreContextLock.RLock()
+	defer c.coreContextLock.RUnlock()
+	foreignChains := make([]chains.Chain, 0)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/core_context/zeta_core_context.go:93]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: zetaclient/zetacore_observer.go [hunk 9]
```diff
@@ -148,25 +150,74 @@ func (co *CoreObserver) GetUpdatedChainClient(coreContext *corecontext.ZetaCoreC
 	return chainOb, nil
 }
 
-// startCctxScheduler schedules keysigns for cctxs on each ZetaChain block (the ticker)
-func (co *CoreObserver) startCctxScheduler(appContext *appcontext.AppContext) {
+// GetPendingCctxsWithinRatelimit get pending cctxs across foreign chains within rate limit
+func (co *CoreObserver) GetPendingCctxsWithinRatelimit(foreignChains []chains.Chain, logger zerolog.Logger) (map[int64][]*types.CrossChainTx, error) {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go]: https://github.com/zeta-chain/node/pull/2110/commits/77acc5bc75273b08ec6091f050c38e74e78e4ddc

File: zetaclient/zetacore_observer.go [hunk 10]
```diff
@@ -148,25 +150,74 @@ func (co *CoreObserver) GetUpdatedChainClient(coreContext *corecontext.ZetaCoreC
...
+	// get rate limiter flags
+	rateLimitFlags, err := co.bridge.GetRateLimiterFlags()
+	if err != nil {
+		return nil, err
+	}
+
+	// apply rate limiter or not according to the flags
+	applyLimit := ratelimiter.IsRateLimiterUsable(rateLimitFlags)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go]: https://github.com/zeta-chain/node/pull/2110/commits/fab00cdc7cfa270fa4f4c28dec1e9f6e5a80e099

File: zetaclient/zetacore_observer.go [hunk 11]
```diff
@@ -253,41 +298,17 @@ func (co *CoreObserver) startCctxScheduler(appContext *appcontext.AppContext) {
...
+// ScheduleCctxEVM schedules evm outtx keysign on each ZetaChain block (the ticker)
+func (co *CoreObserver) ScheduleCctxEVM(
 	outTxMan *outtxprocessor.Processor,
 	zetaHeight uint64,
 	chainID int64,
 	cctxList []*types.CrossChainTx,
 	ob interfaces.ChainClient,
-	signer interfaces.ChainSigner,
-) {
+	signer interfaces.ChainSigner) {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go]: https://github.com/zeta-chain/node/pull/2110/commits/fab00cdc7cfa270fa4f4c28dec1e9f6e5a80e099

**Review by lumtis — APPROVED** [review]

File: x/crosschain/keeper/grpc_query_cctx_rate_limit.go [hunk 12]
```diff
@@ -3,18 +3,172 @@ package keeper
...
+	}
+
+	// the `limit` of pending result is reached or not
+	maxCCTXsReached := func(cctxs []*types.CrossChainTx) bool {
+		// #nosec G701 len always positive
+		return uint32(len(cctxs)) > limit
+	}
+
+	// if a cctx falls within the rate limiter window
+	isCctxInWindow := func(cctx *types.CrossChainTx) bool {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx_rate_limit.go:54]: The other definitions will be removed once we switch to this refactored version in next release. They're all one-time-use functions.

File: x/crosschain/types/rate_limiter_flags.go [hunk 13]
```diff
@@ -45,3 +48,95 @@ func (r RateLimiterFlags) GetConversionRate(zrc20 string) (sdk.Dec, bool) {
...
+		}
+	}
+	return gasAssetRateMap, erc20AssetRateMap
+}
+
+// ConvertCctxValue converts the value of the cctx to azeta using given conversion rates
+func ConvertCctxValue(
+	chainID int64,
+	cctx *CrossChainTx,
+	gasAssetRateMap map[int64]AssetRate,
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/types/rate_limiter_flags.go:79]: Love this.

File: pkg/math/float.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// Percentage calculates the percentage of A over B.
+func Percentage(a, b *big.Int) *big.Float {
+	// if a is zero, return nil to avoid division by zero
+	if b.Cmp(big.NewInt(0)) == 0 {
+		return nil
+	}
+
+	// convert a and a to big.Float
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply pkg/math/float.go]: https://github.com/zeta-chain/node/pull/2110/commits/f63f26488a1b9daacb8bed020c01880451458c0d

File: x/crosschain/types/rate_limiter_flags.go [hunk 15]
```diff
@@ -45,3 +48,95 @@ func (r RateLimiterFlags) GetConversionRate(zrc20 string) (sdk.Dec, bool) {
...
+				erc20AssetRateMap[chainID] = make(map[string]AssetRate)
+			}
+			erc20AssetRateMap[chainID][strings.ToLower(assetRate.Asset)] = assetRate
+		}
+	}
+	return gasAssetRateMap, erc20AssetRateMap
+}
+
+// ConvertCctxValue converts the value of the cctx to azeta using given conversion rates
+func ConvertCctxValue(
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/types/rate_limiter_flags.go]: https://github.com/zeta-chain/node/pull/2110/commits/f63f26488a1b9daacb8bed020c01880451458c0d

File: x/crosschain/keeper/grpc_query_cctx_rate_limit.go [hunk 16]
```diff
@@ -3,18 +3,172 @@ package keeper
...
-	"github.com/zeta-chain/zetacore/pkg/coin"
 	"github.com/zeta-chain/zetacore/x/crosschain/types"
-	fungibletypes "github.com/zeta-chain/zetacore/x/fungible/types"
 	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
+// RateLimiterInput collects the input data for the rate limiter
+func (k Keeper) RateLimiterInput(c context.Context, req *types.QueryRateLimiterInputRequest) (res *types.QueryRateLimiterInputResponse, err error) {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply x/crosschain/keeper/grpc_query_cctx_rate_limit.go:16]: Got what you meant. The 1st loop does the job of `GetPendingNoncesAndFindLowestPendingCctxHeight`. Refactor would involve touching many of existing test cases. I would leave it as is and do the refactor until current rate limiter impl needs improvement or tweak in the future.

**Pull request merged**